### ISSUE TITLE
Version 5 packages discussion

### ISSUE THREAD

**Issue by thomasthesecond** [lonelyplanet/backpack-ui#617]
I've been making some progress splitting up backpack-ui into multiple packages, which will be handled by lerna. Up until this point, this is the folder structure I was considering:

```
backpack-ui
  packages
    backpack-ui-components
      avatar
      button
      heading
      ...etc
    backpack-ui-global-css
    backpack-ui-icons
    backpack-ui-styles
```

Everything related to the UI would be contained in a single monorepo, backpack-ui.

However, I've been thinking about making each individual *component* its own package, which might look like this:

```
backpack-ui-components
  packages
    avatar
    button
    heading
    ...etc

backpack-ui-global-css

backpack-ui-icons

backpack-ui-styles
```

The idea here is that only `backpack-ui-components` is managed by lerna and each component would be installed as needed. `global-css`, `icons` and `styles`would be separate repos. I'm starting to see more component libraries handled this way. The more I think about it, the more I like the idea.

This would allow us to keep `backpack-ui` in it's current state at version 4, but it would be deprecated, similar to what Babel did https://yarnpkg.com/en/package/babel. Doing this would also allow us to keep https://lonelyplanet.github.io/backpack-ui around; I'm not sure how we keep this if  we go with the first monorepo approach, but this isn't a huge issue.

The only thing I don't like about the second approach is there are four separate repos to manage, but `global-css`, `icons` and `styles` won't be very big packages and I don't see them being updated as frequent as `components`.

Thoughts?

**Comment by nickfranciosi** [comment]
@thomasthesecond I like the idea of each component being its own package. Not sure how this is implemented, but it might be worth looking into the way lodash does it. You are able to `npm install lodash` or `npm install lodash.map` to pull in either the entire library or just a piece. 

So I am understanding correctly that `backpack-ui-components` would just have markup/logic and all styles would be pulled in via `backpack-ui-styles`?

**Comment by jcreamer898** [comment]
lodash's build process is insanity

**Comment by thomasthesecond** [comment]
@nickfranciosi Yeah good idea, I think it would be cool to be able to load all of the components at once if you wanted.

Yeah `component` would be all the react code, and the `styles` would be colors, fonts, etc and be imported similar to travel-agent https://github.com/lonelyplanet/travel-agent/blob/master/packages/travel-agent-server/src/webpack/loaders.ts#L9-L18

I found this https://github.com/cloudflare/cf-ui/tree/master/packages and they have components, styles and utils in a single monorepo, they just namespace each package. Maybe that's an option for us?

```
backpack-ui
  packages
    lp-component-button
    lp-component-heading
    lp-icon-close
    lp-icon-share
    lp-style-colors
    lp-style-fonts
```

**Comment by shawncothran** [comment]
I initially like the idea of each component being its own package, but I think loading everything at once is a big plus too. Is that reasonable, seeing as 
> lodash's build process is insanity
? Is that in the scope of Lerna, or would it require something else?
Having `global-css`, `icons` and `styles` separate would be great. I agree that they are touched far less frequently than components, and probably should be touched even less that they are as we stick to our style conventions and guidelines. Having a more atomic approach to components will help in that too.

**Comment by lburdock** [comment]
I'm kind of liking Cloudflare's [cf-ui](https://github.com/cloudflare/cf-ui/tree/master/packages) with everything in the same repo, but namespaced. They've also included utility functions as individual packages. I don't know if that's something we want to do to backpack, but if we follow Cloudflare's pattern we could use the same repo instead of creating a new repo for global utils.

**Comment by shawncothran** [comment]
o0o! global utils is certainly useful, does that fit within backpack's (potential) scope?

**Comment by thomasthesecond** [comment]
@shawncothran I agree about having the ability to load everything, and initially I was thinking that we could output a components.css file for partner sites to use. I'm not sure we could do this as easily with the component-as-separate package approach.

@lburdock It is an interesting approach. For our backpack-ui utils, I figured we would want to create a new package for all of those, but when I actually looked at what our utils were, they are mostly style transformation utils for radium (like mixins) which we won't need going forward with CSS modules. But I can see us having utils that are more useful to the components themselves.

Or maybe the utils aren't scoped to backpack-ui? I could see us having a separate repo with every reusable utility we have across our apps, something like `dotcom-utils`. And then any React app could import the utils they need.

**Comment by thomasthesecond** [comment]
I also just realized that cf-ui is no longer maintained according to their readme 😄

**Comment by thomasthesecond** [comment]
https://github.com/pluralsight/design-system/tree/master/packages is another nice one. They have a `core` package which seems to contain their styles.

**Comment by shawncothran** [comment]
I really like the idea of `dotcom-utils`

**Comment by jcreamer898** [comment]
A big part of this discussion needs to be how to do our styleguide as well. Whether we roll our own with react-docgen, use Bluekit, or X tool.

**Comment by jcreamer898** [comment]
Let's take a step back and create a list of problems we're actually trying to solve with this refactor @thomasthesecond.

**Comment by jcreamer898** [comment]
BTW, this is kinda nice... https://github.com/pluralsight/design-system
